### PR TITLE
[prometheusremotewriteexporter] dropping the condition to replace _ with key_ as __ label is reserved and _ is not

### PR DESCRIPTION
**Pull request by saikrishna397**
__ is a reserved label and _ is not. Being able to use or not use a certain kind of naming convention would be a decision at the exporter level as the nature of the storage dictates what can and can’t be used. it would be nice if we could ideally relax this check for processors in general and backends that don't mandate it like prometheus.

Description: dropped the check to handle labels that start with '_'

Testing: emit a metrics that has a label starting with _ and it should not be prefixed with key

### CODE REVIEW

**Review by Aneurysm9 — DISMISSED** [review]
This will change the behavior of this exporter and possibly break user dashboards. I think this needs a strong use case to justify it. I also still don't know that we understand why it was originally deemed necessary and thus I'm reluctant to remove it without clear demonstration that it is not needed and that removing it will have no harmful effects.

**Comment by vjsamuel** [comment]
@Aneurysm9, i cant speak for why the original author designed this exporter the way they did but going by the name of `prometheusremotewriteexporter`, there is no portion of the prometheus spec that requires masking of labels that start with `_`. the only clause that the prometheus spec has is to dictate the use of `__` for management purposes alone as per:
https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels

Given that prometheus has remote write apis that allow users to directly ingest timeseries via the remote write protocol, i see that they also dont do any such sanitization as per:
https://github.com/prometheus/prometheus/blob/main/storage/remote/codec.go#L485-L495

that being said, the exporter should abide by the spec right?

historically i see that opencensus had this sanitization in place and my assumption is that this piece of code got carried over into various processors and exporters via copy/paste. i can not wholly justify that this change will not break/have no harmful effects but i hope that the above is sufficient to justify that we are doing something extra that is not in the spec that may prevent users to migrate over into using open-telemetry-collector.

**Comment by vjsamuel** [comment]
@Aneurysm9 how do we move forward with this one? can we bare minimum make this configurable with the current behavior set to the default so that it doesnt break anyone's existing functionalities but we provide a way to replicate stock behavior of Prometheus?

**Comment by Aneurysm9** [comment]
> @Aneurysm9 how do we move forward with this one? can we bare minimum make this configurable with the current behavior set to the default so that it doesnt break anyone's existing functionalities but we provide a way to replicate stock behavior of Prometheus?

Please add a [feature gate](https://github.com/open-telemetry/opentelemetry-collector/tree/main/service/featuregate) that can be used to control this behavior.  It should start defaulted to disabled and we can adjust it later.  

See [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/prometheusreceiver/config.go#L43-L47) for an example of a gate in use.  A gate for this component should probably have an ID starting with `exporter.prometheusremotewrite.`.

File: exporter/prometheusremotewriteexporter/exporter.go [hunk 1]
```diff
@@ -180,18 +182,27 @@ func (prwe *PRWExporter) PushMetrics(ctx context.Context, md pdata.Metrics) erro
 	}
 }
 
-func validateAndSanitizeExternalLabels(externalLabels map[string]string) (map[string]string, error) {
+func validateAndSanitizeExternalLabels(externalLabels map[string]string, sanitizeLabel bool) (map[string]string, error) {
```
Comment (Aneurysm9) [inline exporter/prometheusremotewriteexporter/exporter.go:185]:
```suggestion
func validateAndSanitizeExternalLabels(cfg *Config) (map[string]string, error) {
```

The `Config` struct has both the external labels and the sanitization flag.  Passing it here means there is no need to change the signature again when the feature gate is removed.

File: exporter/prometheusremotewriteexporter/config.go [hunk 2]
```diff
@@ -64,6 +66,16 @@ type RemoteWriteQueue struct {
 	NumConsumers int `mapstructure:"num_consumers"`
 }
 
+var dropSanitizationGate = featuregate.Gate{
+	ID:          "exporter.prometheusremotewrite.SanitizeLabel",
+	Enabled:     true,
```
Comment (Aneurysm9) [inline exporter/prometheusremotewriteexporter/config.go:71]:
```suggestion
	ID:          "exporter.prometheusremotewrite.PermissiveLabelSanitizatin",
	Enabled:     false,
```

Feature gates should start disabled by default, eventually transition to enabled by default if the feature is working well, and then finally be removed entirely when the feature is always enabled and users can no longer opt out.

File: exporter/prometheusremotewriteexporter/exporter.go [hunk 3]
```diff
@@ -180,18 +182,27 @@ func (prwe *PRWExporter) PushMetrics(ctx context.Context, md pdata.Metrics) erro
...
 
 		// Sanitize label keys to meet Prometheus Requirements
+		// if sanitizeLabel is enabled, invoke sanitizeLabels else sanitize
 		if len(key) > 2 && key[:2] == "__" {
-			key = "__" + sanitize(key[2:])
+			if sanitizeLabel {
+				key = "__" + sanitizeLabels(key[2:])
+			} else {
+				key = "__" + sanitize(key[2:])
+			}
```
[the remark this answers is not in the document]
Comment (saikrishna397) [reply exporter/prometheusremotewriteexporter/exporter.go:199]: @Aneurysm9  but sanitize() function gets used in many places in helper.go. I I've to pass sanitizeLabel as an argument to sanitize(), it'll need more changes in helper.go.

File: exporter/prometheusremotewriteexporter/exporter.go [hunk 4]
```diff
@@ -50,11 +50,12 @@ type PRWExporter struct {
 	concurrency     int
 	userAgentHeader string
 	clientSettings  *confighttp.HTTPClientSettings
+	sanitizeLabel   bool
 }
 
 // NewPRWExporter initializes a new PRWExporter instance and sets fields accordingly.
-func NewPRWExporter(cfg *Config, buildInfo component.BuildInfo) (*PRWExporter, error) {
-	sanitizedLabels, err := validateAndSanitizeExternalLabels(cfg.ExternalLabels)
+func NewPRWExporter(cfg *Config, buildInfo component.BuildInfo, sanitizeLabel bool) (*PRWExporter, error) {
```
[the remark this answers is not in the document]
Comment (saikrishna397) [reply exporter/prometheusremotewriteexporter/exporter.go]: updated

File: exporter/prometheusremotewriteexporter/exporter.go (hunk 1 above)
Comment (saikrishna397) [reply exporter/prometheusremotewriteexporter/exporter.go]: updated

File: exporter/prometheusremotewriteexporter/config.go (hunk 2 above)
Comment (saikrishna397) [reply exporter/prometheusremotewriteexporter/config.go]: updated

File: exporter/prometheusremotewriteexporter/config.go [hunk 5]
```diff
@@ -64,6 +66,16 @@ type RemoteWriteQueue struct {
...
+	Description: "Controls whether to replace labels starting with _ to key_",
```
[the remark this answers is not in the document]
Comment (saikrishna397) [reply exporter/prometheusremotewriteexporter/config.go]: updated

**Comment by vjsamuel** [comment]
@Aneurysm9, can you please let us know if anything else is pending? if not can we kindly move forward on this one? Sai will file a PR for other prometheus components as discussed on the sync up.

**Review by Aneurysm9 — APPROVED** [review]

File: exporter/prometheusremotewriteexporter/helper.go [hunk 6]
```diff
@@ -102,6 +102,21 @@ func sanitize(s string) string {
...
+func sanitizeLabels(s string) string {
+	if len(s) == 0 {
+		return s
+	}
+
+	// Note: No length limit for label keys because Prometheus doesn't
+	// define a length limit, thus we should NOT be truncating label keys.
+	// labels that start with _ are not sanitized
+	s = strings.Map(sanitizeRune, s)
+	if unicode.IsDigit(rune(s[0])) {
```
Comment (codeboten) [inline exporter/prometheusremotewriteexporter/helper.go:114]: Is this a recommended sanitization for prometheus labels that start w/ a digit? If so, could you add a link to the recommendation here.

File: exporter/prometheusremotewriteexporter/exporter_test.go [hunk 7]
```diff
@@ -685,6 +686,16 @@ func Test_validateAndSanitizeExternalLabels(t *testing.T) {
 			map[string]string{"__key1_key__": "val1"},
 			false,
 		},
+		{"labels_that_start_with_underscore",
+			map[string]string{"_key_": "val1"},
+			map[string]string{"_key_": "val1"},
+			false,
+		},
+		{"labels_that_start_with_digit",
```
Comment (codeboten) [inline exporter/prometheusremotewriteexporter/exporter_test.go:694]: Is there a test to ensure that users with sanitization turned off don't have their labels modified?

File: exporter/prometheusremotewriteexporter/helper.go (hunk 6 above)
Comment (saikrishna397) [reply exporter/prometheusremotewriteexporter/helper.go:114]: https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels

specifies that labels must match the regex [a-zA-Z_:][a-zA-Z0-9_:]* and label names beginning with __ are reserved for internal use.

File: exporter/prometheusremotewriteexporter/exporter_test.go (hunk 7 above)
Comment (saikrishna397) [reply exporter/prometheusremotewriteexporter/exporter_test.go:694]: @codeboten updated the testcases. please review.

**Comment by vjsamuel** [comment]
@codeboten can you please help review this one once again?

**Review by codeboten — APPROVED** [review]

**Pull request merged**